lib/mix/tasks/compile/phoenix_live_view.ex

```diff
@@ -92,53 +92,320 @@ defmodule Mix.Tasks.Compile.PhoenixLiveView do
+                attribute \"#{name}\" \
+                in component #{component(call)} \
+                must be a :list, \
+                got: #{inspect(value)}\
+                """
+
+                [error(message, call.file, line)]
+
+              # struct attribute type mismatch
+              {{:struct, mod}, {line, _column, kind, value}} when kind != :struct ->
+                message = """
+                attribute \"#{name}\" \
+                in component #{component(call)} \
+                must be a #{inspect(mod)}, \
+                got: #{inspect(value)}\
+                """
+
+                [error(message, call.file, line)]
+
+              # attribute type match
+              {_type, {_line, _column, _kind, _value}} ->
+                []
+            end
+
+          {warnings, {attrs, has_global? || type == :global}}
+      end)
 
-              [error(message, call.file, line)]
+    attrs_undefined =
+      for {name, {line, _column, _kind, _value}} <- attrs,
+          not (has_global? and Phoenix.Component.__global__?(caller_module, Atom.to_string(name))) do
+        message = """
+        undefined attribute \"#{name}\" \
+        for component #{component(call)}\
+        """
 
-            {line, _column, string} when is_binary(string) and type not in [:any, :string] ->
-              message =
-                "attribute \"#{name}\" in component #{component(call)} must be a #{inspect(type)}, " <>
-                  "got string: #{inspect(string)}"
+        error(message, call.file, line)
+      end
 
-              [error(message, call.file, line)]
+    {slots_warnings, slots} =
+      Enum.flat_map_reduce(slots_defs, slots, fn slot_def, slots ->
+        %{name: slot_name, required: required} = slot_def
+        slot_attr_defs = get_slot_attr_defs(attrs_defs, slot_name)
+        {slot_values, slots} = Map.pop(slots, slot_name)
 
-            {line, _column, nil} when type not in [:any, :boolean] ->
-              message =
-                "attribute \"#{name}\" in component #{component(call)} must be a #{inspect(type)}, " <>
-                  "got boolean: true"
+        warnings =
+          case {slot_values, slot_attr_defs} do
+            # missing required slot
+            {nil, _slot_attr_defs} when required ->
+              message = """
+              missing required slot \"#{slot_name}\" \
+              for component #{component(call)}\
+              """
 
-              [error(message, call.file, line)]
+              [error(message, call.file, call.line)]
 
-            _ ->
+            # missing optional slot
+            {nil, _slot_attr_defs} ->
               []
+
+            # slot with attributes
+            {slot_values, slot_attr_defs} ->
+              missing_slot_attrs =
+                for slot_value <- slot_values,
+                    {attr_name, %{required: true}} <- slot_attr_defs,
+                    {line, _, _, _} = Map.fetch!(slot_value, :inner_block),
```

Comment:
I think we can have slots without inner blocks, right? Such as:

    <:foo bar="baz" />

Would this line maybe raise then?

Reply:
Based on how the HTMLEngine is implemented, it looks like slots will always have an `:inner_block` assign defined. If it is a self-closing slot, the value of `:inner_block` is `nil`: https://github.com/phoenixframework/phoenix_live_view/blob/0f75ecb8465774eea0ef6c387b17b26601ad33c6/lib/phoenix_live_view/html_engine.ex#L517

This is used to determine the line of the self-closing slot that is missing a required slot attribute.

I suppose I could leverage `:__slot__` instead, but I'm not sure if that makes this any easier to understand 🤔